### ISSUE TITLE
Question about UNAGI Model Reproducibility

### ISSUE THREAD

**Issue by wangpeng905** [mcgilldinglab/UNAGI#22]
Hello,

Thank you for your previous help. 

When the UNAGI model was executed twice on the same dataset with identical parameters, discrepancies in the results were observed. Specifically, during the initial iterations, the cell clustering exhibited identical characteristics in both runs. However, following the initial iteration, a divergence in cell clustering emerged, resulting in disparities in cell progression dynamics markers. 

<img width="477" height="433" alt="Image" src="https://github.com/user-attachments/assets/0937307e-0191-4fef-97da-6b498c6fa57b" />

<img width="477" height="817" alt="Image" src="https://github.com/user-attachments/assets/69b35f4b-afc0-44b8-82ac-088982d08251" />

This suggests the presence of randomness in the iteration process. The question remains whether this randomness can be mitigated. In the event that this is the case, what adjustments to the parameters should be made to enhance consistency between the two runs?

Thank you so much.

**Comment by wangpeng905** [comment]
Hi @Szym29 ,

When I specified the UNAGI model to use only CPU computation, running the same data twice yielded 100% reproducible results. However, when GPU computation was specified, running it twice produced non-reproducible results. Therefore, the issue likely stems from GPU computation having lower precision than CPU, rather than being a problem with the UNAGI model itself.

When using GPU computation, the model training loss gradually diverges as epochs increase, whereas this divergence does not occur when using CPU computation.

```shell
$ paste round8.1/0/loss.txt round8.2/0/loss.txt
[epoch 000]  average training loss: 3070.0751   [epoch 000]  average training loss: 3070.0751
[epoch 001]  average training loss: 2624.7252   [epoch 001]  average training loss: 2624.7252
[epoch 002]  average training loss: 2330.2928   [epoch 002]  average training loss: 2330.2928
[epoch 003]  average training loss: 2189.9965   [epoch 003]  average training loss: 2189.9967
[epoch 004]  average training loss: 2096.9101   [epoch 004]  average training loss: 2096.9103
[epoch 005]  average training loss: 2028.4249   [epoch 005]  average training loss: 2028.4259
[epoch 006]  average training loss: 1974.5283   [epoch 006]  average training loss: 1974.5323
[epoch 007]  average training loss: 1941.9389   [epoch 007]  average training loss: 1941.9427
[epoch 008]  average training loss: 1913.2752   [epoch 008]  average training loss: 1913.2782
[epoch 009]  average training loss: 1895.0940   [epoch 009]  average training loss: 1895.0953
```

round8.1/0/loss.txt and round8.2/0/loss.txt represent the training loss from the first iteration of two GPU-based runs. The divergence becomes visible starting from epoch 003.
